docs/exec_model.md

```diff
@@ -0,0 +1,154 @@
+completion without any kind of preemption. A long running task may "block" the
+execution of other tasks. This is especially noticeable in a Web browser when
+the DOM rendering task and the UI event handler tasks are blocked.
+
+![JS execution model](js_exec_model.svg)
+
+Each task is in fact composed of a queue of "micro-tasks". Initially there is a
+single micro-task in the queue. This task can queue other micro-tasks with
+``queueMicrotask(task)``. These micro-tasks are executed in FIFO order. When the
+micro-task queue is empty, the macro-task has completed.
+
+**Promises**
+
+Promises are JS synchronization objects. When they settle they can either be
+resolved or rejected.
+
+```
+p = new Promise((resolve,reject) => whatever)
+```
+
+``whatever`` is executed synchronously. ``resolve`` and ``reject`` are functions to set the Promise value.
+
+```
+p.then(r => doSomething)
+p.catch(r => doSomething)
+```
+
+``doSomething`` is always executed asynchronously.
+
+In a micro-task we detect if ``p`` is settled: if it is, we execute ``doSomething``,
+otherwise ``doSomething`` will be executed in a future macro-task (when ``p`` gets
+settled)
+
+```
+b = p.then(f).then(g)
+```
+
+is equivalent to:
+
+```
+a = p.then(f)
+b = a.then(g)
+```
+so we queue several micro-tasks in order just like before. If we only have
+settled Promises, the execution is very fast as no macro-task creation/switching
+occurs.
+
+``Promise.all(xs)`` waits for all the promises to settle.
+``Promise.race(xs)`` returns the first settled Promise in xs (eiter resolved or rejected)
+
+With ES2016 we can somewhat replace Promises with await/async:
+
+```
+async function (x,y) {
+   try {
+      let a = await p
+      let b = await (f a)
+      ...
+   } catch(errhandler)
+}
+```
+
+* ``await`` can only be used in an ``async`` function
+* ``await p`` creates a delimited continuation ``c`` so that ``p.then(c)``
+
+**Workers**
+
+JS supports some kind of multi-threading via
+[Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API). A
+Worker has its own task queues and execution engine. Each worker has its own JS
+context and Workers are executed concurrently (e.g. one OS thread per Worker).
+
+Communication between Workers is done via message passing. Using a shared memory
+was possible through
+[SharedBufferArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer)
+but it has been disabled for security reasons.
+
+**Wasm threads**
+
+There is a
+[proposal](https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md)
+to add support for threads in WebAssembly reusing Worker and SharedBufferArray
+but it isn't implemented yet.
+
+## Haskell execution model
+
+Haskell threads are represented by a Thread State Object (TSO). Each TSO has its
+own Stack.
+
+In the native RTS, TSOs are executed by Capabilities (virtual CPUs). Each
```

Comment:
It's worth noting that the `Task` and `InCall` concepts do not exist in our own runtime; we have a single global `MainCapability` just like the non-threaded runtime. Also, we don't implement the "run queue" yet, and even if we do, it's likely we'll do it in JavaScript instead of using the TSO queue objects on the wasm heap.